pkg/cmd/cli/sa/newtoken.go

```diff
@@ -0,0 +1,235 @@
+	}
+
+	namespace, _, err := f.DefaultNamespace()
+	if err != nil {
+		return fmt.Errorf("could not retrieve default namespace: %v", err)
+	}
+
+	o.SAClient = client.ServiceAccounts(namespace)
+	o.SecretsClient = client.Secrets(namespace)
+	return nil
+}
+
+func (o *NewServiceAccountTokenOptions) Validate() error {
+	if o.SAName == "" {
+		return errors.New("service account name cannot be empty")
+	}
+
+	if o.SAClient == nil || o.SecretsClient == nil {
+		return errors.New("API clients must not be nil in order to create a new service account token")
+	}
+
+	if o.Timeout <= 0 {
+		return errors.New("a positive amount of time must be allotted for the timeout")
+	}
+
+	if o.Out == nil || o.Err == nil {
+		return errors.New("cannot proceed if output or error writers are nil")
+	}
+
+	return nil
+}
+
+// Run creates a new token secret, waits for the service account token controller to fulfill it, then adds the token to the service account
+func (o *NewServiceAccountTokenOptions) Run() error {
+	serviceAccount, err := o.SAClient.Get(o.SAName)
+	if err != nil {
+		return err
+	}
+
+	tokenSecret := &api.Secret{
+		ObjectMeta: api.ObjectMeta{
+			GenerateName: osautil.GetTokenSecretNamePrefix(serviceAccount),
+			Namespace:    serviceAccount.Namespace,
+			Labels:       o.Labels,
+			Annotations: map[string]string{
+				api.ServiceAccountNameKey: serviceAccount.Name,
+			},
+		},
+		Type: api.SecretTypeServiceAccountToken,
+		Data: map[string][]byte{},
+	}
+
+	persistedToken, err := o.SecretsClient.Create(tokenSecret)
+	if err != nil {
+		return err
+	}
+
+	// we need to wait for the service account token controller to make the new token valid
+	tokenSecret, err = waitForToken(persistedToken, serviceAccount, o.Timeout, o.SecretsClient)
+	if err != nil {
+		return err
+	}
+
+	token, exists := tokenSecret.Data[api.ServiceAccountTokenKey]
+	if !exists {
+		return fmt.Errorf("service account token %q did not contain token data", tokenSecret.Name)
+	}
+
+	fmt.Fprintf(o.Out, string(token))
+	if util.IsTerminalWriter(o.Out) {
+		// pretty-print for a TTY
+		fmt.Fprintf(o.Out, "\n")
+	}
+	return nil
+}
+
+// waitForToken uses `cmd.Until` to wait for the service account controller to fulfill the token request
+func waitForToken(token *api.Secret, serviceAccount *api.ServiceAccount, timeout time.Duration, client unversioned.SecretsInterface) (*api.Secret, error) {
+	// there is no provided rounding function, so we use Round(x) === Floor(x + 0.5)
+	timeoutSeconds := int64(math.Floor(timeout.Seconds() + 0.5))
+
+	options := api.ListOptions{
+		FieldSelector:   fields.SelectorFromSet(fields.Set(map[string]string{"metadata.name": token.Name})),
+		Watch:           true,
+		ResourceVersion: token.ResourceVersion,
+		TimeoutSeconds:  &timeoutSeconds,
+	}
+
+	watcher, err := client.Watch(options)
+	if err != nil {
+		return nil, fmt.Errorf("could not begin watch for token: %v", err)
+	}
+
+	event, err := cmd.Until(timeout, watcher, func(event watch.Event) (bool, error) {
```

Comment:
nit: I prefer methods bigger than a line or two be declared.  It allows unit testing.